# Multiple Linear Regression with sklearn - Exercise Solution

You are given a real estate dataset. 

Real estate is one of those examples that every regression course goes through as it is extremely easy to understand and there is a (almost always) certain causal relationship to be found.

The data is located in the file: 'real_estate_price_size_year.csv'. 

You are expected to create a multiple linear regression (similar to the one in the lecture), using the new data. 

Apart from that, please:
-  Display the intercept and coefficient(s)
-  Find the R-squared and Adjusted R-squared
-  Compare the R-squared and the Adjusted R-squared
-  Compare the R-squared of this regression and the simple linear regression where only 'size' was used
-  Using the model make a prediction about an apartment with size 750 sq.ft. from 2009
-  Find the univariate (or multivariate if you wish - see the article) p-values of the two variables. What can you say about them?
-  Create a summary table with your findings

In this exercise, the dependent variable is 'price', while the independent variables are 'size' and 'year'.

Good luck!

## Import the relevant libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

#import sklearn

from sklearn.linear_model import LinearRegression

## Load the data

In [2]:
data = pd.read_csv('real_estate_price_size_year.csv')

data.head()

price     size  year
0  234314.144   643.09  2015
1  228581.528   656.22  2009
2  281626.336   487.29  2018
3  401255.608  1504.75  2015
4  458674.256  1275.46  2009

In [3]:
#lets look at descriptive statistics
data.describe()

price         size         year
count     100.000000   100.000000   100.000000
mean   292289.470160   853.024200  2012.600000
std     77051.727525   297.941951     4.729021
min    154282.128000   479.750000  2006.000000
25%    234280.148000   643.330000  2009.000000
50%    280590.716000   696.405000  2015.000000
75%    335723.696000  1029.322500  2018.000000
max    500681.128000  1842.510000  2018.000000

## Create the regression

### Declare the dependent and the independent variables

In [4]:
#independent
x = data[['size','year']]

#dependent
y = data['price']

### Regression

In [5]:
#create linear regression object 
reg = LinearRegression()

#fit regression. Order matters (x,y)
reg.fit(x,y)

LinearRegression()

### Find the intercept

In [6]:
reg.intercept_

-5772267.017463281

### Find the coefficients

In [7]:
reg.coef_

array([ 227.70085401, 2916.78532684])

### Calculate the R-squared

In [8]:
reg.score(x,y)

0.7764803683276793

In [9]:
x.shape

(100, 2)

### Calculate the Adjusted R-squared

$R^2_{adj.} = 1 - (1-R^2) * \frac{n-1}{n-p-1}$

In [10]:
r2 = reg.score(x,y) 

#number of samples aka observations
n = x.shape[0]

#num of features
p = x.shape[1]

adjusted_r2 = 1 - (1-r2) * ((n-1)/(n-p-1))

adjusted_r2

0.77187171612825

### Compare the R-squared and the Adjusted R-squared

Answer...

r2 is **0.7764803683276793** and adjusted_r2 is **0.77187171612825**

r2 is slightly larger meaning that our inclusion of year into the regression model was not really penalized

### Compare the Adjusted R-squared with the R-squared of the simple linear regression

Answer...

'Year' is not bringing too much value to the result

### Making predictions

Find the predicted price of an apartment that has a size of 750 sq.ft. from 2009.

In [11]:
reg.predict([[750,2009]])

/Users/joeeey.d/opt/anaconda3/lib/python3.9/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([258330.34465995])

### Calculate the univariate p-values of the variables

In [12]:
#import feature selection 
from sklearn.feature_selection import f_regression 

In [13]:
#look into f-regression to find f-statistics and p-values
f_regression(x,y)

#first element is f statistics, second element is p-values of each feature in ORDER

(array([285.92105192,   0.85525799]), array([8.12763222e-31, 3.57340758e-01]))

In [14]:
#lets grab the p-values
p_values = f_regression(x,y)[1]
p_values

array([8.12763222e-31, 3.57340758e-01])

In [15]:
#round them
p_values.round(3)

array([0.   , 0.357])

Interesting. Year has a p-value of 0.357, meaning its above the p-value of significance which is **0.05**. We can declare year as insiginificant

### Create a summary table with your findings

In [16]:
# Let's create a new data frame with the names of the features
reg_summary = pd.DataFrame(data = x.columns.values, columns=['Features'])

reg_summary

#_.columns.values gets you the gives you the column names

Features
0     size
1     year

In [17]:
x

size  year
0    643.09  2015
1    656.22  2009
2    487.29  2018
3   1504.75  2015
4   1275.46  2009
..      ...   ...
95   549.80  2009
96  1037.44  2009
97  1504.75  2006
98   648.29  2015
99   705.29  2006

[100 rows x 2 columns]

In [19]:
#add the coefficients to the summary table
reg_summary['Coefficients'] = reg.coef_

reg_summary

Features  Coefficients
0     size    227.700854
1     year   2916.785327

In [20]:
#add p_values
reg_summary['p_values'] = p_values.round(3)
reg_summary

Features  Coefficients  p_values
0     size    227.700854     0.000
1     year   2916.785327     0.357

Answer...

Once again, year is insiginficant. It should be removed. 